src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -41,27 +42,44 @@
   @Override
   /* Show courses. */
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    List<Course> courses = getHelper(request);
-
-    String coursesJson = new Gson().toJson(courses);
+    JSONObject results = getMatchingCourses(request);
     response.setContentType("application/json;");
-    response.getWriter().println(coursesJson);
+    response.getWriter().println(results);
   }
 
   /* Create list of courses given request. Used for testing. */
-  public List<Course> getHelper(HttpServletRequest request) {
-    List<Filter> filters = getFilters(request);
+  public JSONObject getMatchingCourses(HttpServletRequest request) {
+    String success = "good";
+    List<Filter> filters = getFilters(request, false);
     List<Entity> results = getResults(filters);
-    List<Course> courses = getCourses(results);
+    if (results.isEmpty()) {
+      filters = getFilters(request, true); // Try fuzzy search.
+      results = getResults(filters);
+      success = "okay";
+    }
+    if (results.isEmpty()) { // Fuzzy search didn't work.
+      success = "bad";
+    }
+    JSONObject courses = getCourses(results, success);
     return courses;
   }
 
   /* Create list of filters given parameters specified in request. */
-  private List<Filter> getFilters(HttpServletRequest request) {
+  private List<Filter> getFilters(HttpServletRequest request, boolean fuzzy) {
     List<Filter> filters = new ArrayList<>();
     if (!request.getParameter("courseName").isEmpty()) {
       String name = request.getParameter("courseName");
-      Filter nameFilter = new FilterPredicate("name", FilterOperator.EQUAL, name);
+      Filter nameFilter;
+      if (fuzzy) {
+        String dept = name.split(" ")[0];
```

Comment:
Another point where shortening the name doesn't really make sense. I'm guessing this is department, but it could also be depot, or a few other things.
Prefer more descriptive names in general, that way I don't have to remember context to remember what a variable is referring to.

Reply:
Also it removes any second guessing when debugging something. Having in the back of your mind "I assumed it meant department, but what if it didn't?" takes up brain space and can delay a solution.